clippy_lints/src/redundant_test_prefix.rs

```diff
@@ -0,0 +1,74 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::{is_in_cfg_test, is_in_test_function};
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::def_id::LocalDefId;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Triggers when a function annotated with the `#[test]` macro begins with `test_`
+    /// ### Why is this bad?
+    /// This is redundant, since the annotations already denotes the function a test.
+    /// Tests are executed using `cargo test `module::test-name-here::`.
+    /// It's clear that a test is a test, without prefixing `test_` to the test name.
+    /// ### Example
+    /// ```no_run
+    /// #[test]
+    /// fn my_test_case() {
+    ///     // test logic here...
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// #[test]
+    /// fn my_test_case() {
+    ///     // test logic here...
+    /// }
+    /// ```
+    #[clippy::version = "1.79.0"]
+    pub REDUNDANT_TEST_PREFIX,
+    pedantic,
```

Comment:
I feel like this shouldn't be in `pedantic` but in `restriction` instead. The usage of the `test_` prefix doesn't seem like an antipattern to me. There are good arguments against it, but the number of changes required in existing changes shows that the usage is still quite common.

I'll include this decision in the group discussion before we merge this PR :)

Reply:
There are arguments to be made for several categories: `style`, `pedantic`, `restriction`, etc.

The usage of `test_` seems to be `pedantic` to me, because it's clearly unnecessary in almost every case, including some of the ones mentioned in the threads in this PR. For example, test names that have `test_` + `method_name`, should just be `super::method_name` IMO, following the same like of reasoning for as everywhere else.

Reply:
One argument I always consider is the number of lint triggers that happen in current code. Clippy can be a very pedantic, but the number of lint emissions has to be reasonable. If users get 100+ warnings of this lint with the next Clippy release it's way more likely that users get annoyed and simply allow the lint and that would help nobody.

This is not meant as a hard no, but just an explanation of my suggestion. I'm not sure where to place it and would leave this up to a group decision.